<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Hich Reference &#8212; Hich 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=c058f7c8" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=2709fde1"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Hich Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="hich-reference">
<h1>Hich Reference<a class="headerlink" href="#hich-reference" title="Link to this heading">¶</a></h1>
<p>In Hich, a <strong>sample</strong> is a single unit of data, such as a <strong>technical replicate</strong>, <strong>biological replicate</strong>, or experimental <strong>condition</strong>.</p>
<section id="sample-attributes">
<h2>Sample attributes<a class="headerlink" href="#sample-attributes" title="Link to this heading">¶</a></h2>
<section id="general">
<h3>General<a class="headerlink" href="#general" title="Link to this heading">¶</a></h3>
<section id="sample-id">
<h4>sample_id<a class="headerlink" href="#sample-id" title="Link to this heading">¶</a></h4>
<p>Required (no default)</p>
<p>A unique name for the sample.</p>
</section>
<section id="id">
<h4>id<a class="headerlink" href="#id" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required (defaults to sample_id)</div>
<div class="line"><br /></div>
<div class="line">A unique name for the sample. Automatically chosen by Hich, not set manually.</div>
</div>
</section>
<section id="assembly">
<h4>assembly<a class="headerlink" href="#assembly" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required (no default)</div>
<div class="line"><br /></div>
<div class="line">The genome assembly for the sample, such as <code class="docutils literal notranslate"><span class="pre">hg38</span></code>. Hich will automatically download a genome reference if not provided for the following assemblies:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">hg38</span></code>, <code class="docutils literal notranslate"><span class="pre">homo_sapiens</span></code>, <code class="docutils literal notranslate"><span class="pre">GRCh38</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mm10</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dm6</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">galGal5</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bGalGal5</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">danRer11</span></code></p></li>
</ul>
</section>
<section id="datatype">
<h4>datatype<a class="headerlink" href="#datatype" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required (no default)</div>
<div class="line"><br /></div>
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sam</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bam</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pairs</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">The format for input read data.</div>
<div class="line"><br /></div>
<div class="line"><cite>Note: Hich can read data compressed in gzip format, but gzip compression does not need to be explicitly specified.</cite></div>
</div>
</section>
<section id="is-techrep">
<h4>is_techrep<a class="headerlink" href="#is-techrep" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Available options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code> <cite>default</cite></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether or not the sample is a technical replicate.</div>
</div>
</section>
<section id="is-biorep">
<h4>is_biorep<a class="headerlink" href="#is-biorep" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Available options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code> <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether or not the sample is a biological replicate.</div>
</div>
</section>
<section id="is-condition">
<h4>is_condition<a class="headerlink" href="#is-condition" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Available options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code> <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether or not the sample is a condition.</div>
</div>
</section>
<section id="min-mapq">
<h4>min_mapq<a class="headerlink" href="#min-mapq" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Not required</div>
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">30</span></code></div>
<div class="line"><br /></div>
<div class="line">Minimum alignment threshold to keep an aligned read.</div>
</div>
</section>
</section>
<section id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Link to this heading">¶</a></h3>
<section id="aligner">
<h4>aligner<a class="headerlink" href="#aligner" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Required if <code class="docutils literal notranslate"><span class="pre">datatype</span> <span class="pre">==</span> <span class="pre">fastq</span></code></div>
<div class="line">Available options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bwa</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bwa-mem2</span></code> <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">While <cite>bwa-mem2</cite> is 1.3-3x faster, indexing genomes with <code class="docutils literal notranslate"><span class="pre">bwa-mem2</span></code> requires a 60-80 Gb memory footprint, whereas indexing with <cite>bwa</cite> can be done in less than 32 Gb.</div>
</div>
</section>
<section id="aligner-threads">
<h4>aligner_threads<a class="headerlink" href="#aligner-threads" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">10</span></code></div>
<div class="line"><br /></div>
<div class="line">Max threads to use for alignment. It is highly recommended to set this to the maximum number of available cores. Note that only one alignment process is spawned at a time. This is because the aligners Hich uses (BWA MEM and BWA MEM2) are internally parallelized, so there is no substantial performance gain to running multiple alignment processes in parallel, while the substantial memory footprint is duplicated for each aligner instance being run.</div>
</div>
</section>
<section id="bwa-flags">
<h4>bwa_flags<a class="headerlink" href="#bwa-flags" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">-SP5M</span></code></div>
<div class="line"><br /></div>
<div class="line">Flags to use for <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code> or <code class="docutils literal notranslate"><span class="pre">bwa-mem2</span> <span class="pre">mem</span></code>. The default <code class="docutils literal notranslate"><span class="pre">-SP5M</span></code> is recommended by <a class="reference external" href="https://data.4dnucleome.org/resources/data-analysis/hi_c-processing-pipeline">4DN</a> for aligning paired-end Hi-C reads with <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code> or <code class="docutils literal notranslate"><span class="pre">bwa-mem2</span> <span class="pre">mem</span></code>. See <a class="reference external" href="https://bio-bwa.sourceforge.net/bwa.shtml">bwa manual reference page</a> for additional options.</div>
</div>
</section>
</section>
<section id="pairs-processing">
<h3>Pairs processing<a class="headerlink" href="#pairs-processing" title="Link to this heading">¶</a></h3>
<section id="deduplicate">
<h4>deduplicate<a class="headerlink" href="#deduplicate" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>true <cite>default</cite></p></li>
<li><p>false</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to remove technical duplicates (i.e. PCR or optical duplicates). Deduplication is applied to biological replicates after forming them from non-deduplicated technical replicates or after ingesting them directly into Hich. Hich deduplicates technical replicates <cite>after</cite> using them to merge biological replicates.</div>
</div>
</section>
<section id="pairs-format">
<h4>pairs_format<a class="headerlink" href="#pairs-format" title="Link to this heading">¶</a></h4>
<section id="chrom1">
<h5>chrom1<a class="headerlink" href="#chrom1" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Default: 2</div>
<div class="line"><br /></div>
<div class="line">The column in the .pairs file where the first chromosome is labeled for each read.</div>
</div>
</section>
<section id="pos1">
<h5>pos1<a class="headerlink" href="#pos1" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Default: 3</div>
<div class="line"><br /></div>
<div class="line">The column in the .pairs file where the first base pair position is labeled for each read.</div>
</div>
</section>
<section id="chrom2">
<h5>chrom2<a class="headerlink" href="#chrom2" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Default: 4</div>
<div class="line"><br /></div>
<div class="line">The column in the .pairs file where the second chromosome is labeled for each read.</div>
</div>
</section>
<section id="pos2">
<h5>pos2<a class="headerlink" href="#pos2" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Required</div>
<div class="line">Default: 5</div>
<div class="line"><br /></div>
<div class="line">The column in the .pairs file where the second base pair position is labeled for each read.</div>
</div>
</section>
</section>
<section id="parse-params">
<h4>parse_params<a class="headerlink" href="#parse-params" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Default:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--flip</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--drop-readid</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--drop-seq</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--drop-sam</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Extra parameters to use for parsing .sam/.bam alignments into .pairs format.</div>
<div class="line"><br /></div>
<div class="line"><cite>Note:</cite> The <code class="docutils literal notranslate"><span class="pre">drop-*</span></code> parameters are one of the most impactful for making Hich fast and giving it a low disk footprint. It is not recommended to remove these parameters unless you know what you are doing, although additional parameters can be added.</div>
</div>
</section>
<section id="dedup-params">
<h4>dedup_params<a class="headerlink" href="#dedup-params" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Extra parameters to use during the deduplication step.</div>
</div>
</section>
<section id="select-params">
<h4>select_params<a class="headerlink" href="#select-params" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Extra parameters to use during the selection step.</div>
</div>
</section>
<section id="select-condition">
<h4>select_condition<a class="headerlink" href="#select-condition" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Read-level filters to use during the selection step.</div>
</div>
<section id="keep-pair-types">
<h5>keep_pair_types<a class="headerlink" href="#keep-pair-types" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">UU</span></code>, <code class="docutils literal notranslate"><span class="pre">UR</span></code>, <code class="docutils literal notranslate"><span class="pre">RU</span></code></div>
<div class="line"><br /></div>
<div class="line">U is for a unique aligned read, whereas an R is “rescued” by detecting pairs where one side maps to locus 1 and the other to a slightly different position on locus 1 and to locus 2, the classic “split ligation junction” pattern that represents an observed, rather than inferred, ligation junction.</div>
</div>
</section>
<section id="keep-trans">
<h5>keep_trans<a class="headerlink" href="#keep-trans" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code> <cite>default</cite></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to keep interchromosomal (“trans”) contacts. Note that this should be left as true if forming .mcool files and using the default <code class="docutils literal notranslate"><span class="pre">trans-only</span></code> option, which normalizes contact matrices based exclusively on trans contacts, which are in some cases thought to yield more biologically representative results.</div>
</div>
</section>
<section id="keep-cis">
<h5>keep_cis<a class="headerlink" href="#keep-cis" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code> <cite>default</cite></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to keep intrachromosomal (“cis”) contacts.</div>
</div>
</section>
<section id="min-dist-fr">
<h5>min_dist_fr<a class="headerlink" href="#min-dist-fr" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">1000</span></code></div>
<div class="line">Minimum insert size (in bp) to keep FR (+- or inward) strands. In Hi-C, the set of short-range FR strands can be highly enriched in undigested chromatin, which shows up in Hich’s MultiQC report as a percentage of FR orientations substantially higher than the expected 25%. These can be filtered out using this option.</div>
</div>
</section>
<section id="min-dist-rf">
<h5>min_dist_rf<a class="headerlink" href="#min-dist-rf" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">1000</span></code></div>
<div class="line">Minimum insert size (in bp) to keep RF (-+ or outward) strands. In Hi-C, the set of short-range FR strands can be highly enriched in self-circles (digested fragments that self-ligated end to end), which shows up in Hich’s MultiQC report as a percentage of RF orientations substantially higher than the expected 25%. These can be filtered out using this option.</div>
</div>
</section>
<section id="min-dist-ff">
<h5>min_dist_ff<a class="headerlink" href="#min-dist-ff" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">0</span></code></div>
<div class="line">Minimum insert size (in bp) to keep FF (++) strands.</div>
</div>
</section>
<section id="id1">
<h5>min_dist_ff<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: <code class="docutils literal notranslate"><span class="pre">0</span></code></div>
<div class="line">Minimum insert size (in bp) to keep RR (–) strands.</div>
</div>
</section>
<section id="chroms">
<h5>chroms<a class="headerlink" href="#chroms" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">If specified, each read alignment must be to a chromosome in this set.</div>
</div>
</section>
<section id="discard-same-frag">
<h5>discard_same_frag<a class="headerlink" href="#discard-same-frag" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">true</span></code> <cite>default</cite></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">false</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">If true, fragments whose alignments are mapped to restriction fragments will be discarded if both ends mapped to the same restriction fragment.</div>
</div>
</section>
</section>
</section>
<section id="matrix-processing">
<h3>Matrix processing<a class="headerlink" href="#matrix-processing" title="Link to this heading">¶</a></h3>
<section id="make-hic">
<h4>make_hic<a class="headerlink" href="#make-hic" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Arguments supplied to juicer tools’ <code class="docutils literal notranslate"><span class="pre">pre</span></code> command when forming a Hi-C contact matrix.</div>
</div>
</section>
<section id="make-cool">
<h4>make_cool<a class="headerlink" href="#make-cool" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Arguments supplied to the <code class="docutils literal notranslate"><span class="pre">cooler</span> <span class="pre">cload</span></code> command for forming .cool format precursors to the .mcool contact matrix.</div>
</div>
</section>
<section id="make-mcool">
<h4>make_mcool<a class="headerlink" href="#make-mcool" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Default:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--balance</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--balance-args</span> <span class="pre">'max-iters</span> <span class="pre">2000</span> <span class="pre">--trans-only'</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Arguments supplied to the <code class="docutils literal notranslate"><span class="pre">cooler</span> <span class="pre">zoomify</span></code> command for coarsening high-res .cool matrices into multi-resolution .mcool contact matrices. The chosen defaults will generate multi-res contact matrices containing both the raw contacts and balancing weights produced using the trans contacts only.</div>
</div>
</section>
<section id="matrix">
<h4>matrix<a class="headerlink" href="#matrix" title="Link to this heading">¶</a></h4>
<section id="make-mcool-file-format">
<h5>make_mcool_file_format<a class="headerlink" href="#make-mcool-file-format" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>true <cite>default</cite></p></li>
<li><p>false</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to produce .mcool-format contact matrices (the Open2C multi-resolution format). Currently required for feature calling and QC.</div>
</div>
</section>
<section id="make-hic-file-format">
<h5>make_hic_file_format<a class="headerlink" href="#make-hic-file-format" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>true <cite>default</cite></p></li>
<li><p>false <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to produce .hic-format contact matrices (compatible with the Juicer tool ecosystem including the Juicebox browser).</div>
</div>
</section>
<section id="resolutions">
<h5>resolutions<a class="headerlink" href="#resolutions" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default:</div>
</div>
<ul class="simple">
<li><p>1000</p></li>
<li><p>2000</p></li>
<li><p>5000</p></li>
<li><p>10000</p></li>
<li><p>20000</p></li>
<li><p>50000</p></li>
<li><p>100000</p></li>
<li><p>200000</p></li>
<li><p>500000</p></li>
<li><p>1000000</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Reference chromosome coordinates will be partitioned into these uniform block sizes (in bp) and contact ends mapped to those blocks to generate contact matrices. Lower numbers represent higher-resolution matrices.</div>
</div>
</section>
</section>
</section>
<section id="quality-control">
<h3>Quality control<a class="headerlink" href="#quality-control" title="Link to this heading">¶</a></h3>
<section id="hicrep">
<h4>hicrep<a class="headerlink" href="#hicrep" title="Link to this heading">¶</a></h4>
<section id="call-on">
<h5>call_on<a class="headerlink" href="#call-on" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>is_techrep <cite>default</cite></p></li>
<li><p>is_biorep <cite>default</cite></p></li>
<li><p>is_condition <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to compute Hicrep SCC scores on technical replicates, biological replicates, and conditions. Results for all comparisons are output to a single .tsv file with a per-column header giving the pair of samples, chromosome, resolution, and Hicrep parameters that were used, along with the SCC score.</div>
</div>
</section>
<section id="id2">
<h5>resolutions<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default:</div>
</div>
<ul class="simple">
<li><p>10000</p></li>
<li><p>100000</p></li>
<li><p>1000000</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Which resolutions to use for calling Hicrep SCC scores.</div>
</div>
</section>
<section id="id3">
<h5>chroms<a class="headerlink" href="#id3" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Which chromosomes to use for calling Hicrep SCC scores. If not specified, all chromosomes shared by both matrices at the given resolution will be used.</div>
</div>
</section>
<section id="exclude">
<h5>exclude<a class="headerlink" href="#exclude" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Which chromosomes to exclude when calling Hicrep SCC scores.</div>
</div>
</section>
<section id="chrom-filter">
<h5>chrom_filter<a class="headerlink" href="#chrom-filter" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">A conditional statement in Python to determine whether to use a chromosome for Hicrep as a function of its name (referenced via the <code class="docutils literal notranslate"><span class="pre">chrom</span></code> variable) and size (the <code class="docutils literal notranslate"><span class="pre">size</span></code> variable). It will be evaluated using Python’s <code class="docutils literal notranslate"><span class="pre">eval</span></code> statement.</div>
</div>
</section>
<section id="h">
<h5>h<a class="headerlink" href="#h" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Values of Hicrep’s <code class="docutils literal notranslate"><span class="pre">h</span></code> parameter to use.</div>
</div>
</section>
<section id="dbpmax">
<h5>dBPMax<a class="headerlink" href="#dbpmax" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Values of Hicrep’s <code class="docutils literal notranslate"><span class="pre">dBPMax</span></code> parameter to use.</div>
</div>
</section>
<section id="bdownsample">
<h5>bDownSample<a class="headerlink" href="#bdownsample" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Values of Hicrep’s <code class="docutils literal notranslate"><span class="pre">bDownSample</span></code> parameter to use.</div>
</div>
</section>
</section>
</section>
<section id="feature-calling">
<h3>Feature calling<a class="headerlink" href="#feature-calling" title="Link to this heading">¶</a></h3>
<section id="compartments">
<h4>compartments<a class="headerlink" href="#compartments" title="Link to this heading">¶</a></h4>
<section id="resolution">
<h5>resolution<a class="headerlink" href="#resolution" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: 5000</div>
<div class="line"><br /></div>
<div class="line">The resolution at which compartments should be called.</div>
</div>
</section>
<section id="cooltools-eigs-cis-params">
<h5>cooltools_eigs_cis_params<a class="headerlink" href="#cooltools-eigs-cis-params" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Defaults:</div>
</div>
<ul class="simple">
<li><p>–bigwig</p></li>
</ul>
<div class="line-block">
<div class="line">Additional parameters that should be passed to cooltools_eigs_cis. The default specifies that a .bigwig-format file should be generated as well as the .bedgraph format.</div>
</div>
</section>
</section>
<section id="insulation">
<h4>insulation<a class="headerlink" href="#insulation" title="Link to this heading">¶</a></h4>
<section id="id4">
<h5>resolution<a class="headerlink" href="#id4" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default: 5000</div>
<div class="line"><br /></div>
<div class="line">The resolution at which insulation should be called.</div>
</div>
</section>
<section id="cooltools-insulation-params">
<h5>cooltools_insulation_params<a class="headerlink" href="#cooltools-insulation-params" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Defaults:</div>
</div>
<ul class="simple">
<li><p>–bigwig</p></li>
</ul>
<div class="line-block">
<div class="line">Additional parameters that should be passed to cooltools_eigs_cis. The default specifies that a .bigwig-format file should be generated as well as the .bedgraph format.</div>
</div>
</section>
</section>
<section id="loops">
<h4>loops<a class="headerlink" href="#loops" title="Link to this heading">¶</a></h4>
<div class="line-block">
<div class="line">Hich uses Mustache for loop and differential loop calling. This software was chosen mainly for its theoretical advantages. Based on scale space theory, it applies an artifact-free filter to a matrix to remove fine details, then detects blobs which are called as loops. It thereby takes local information into account in loop calling. Differential loops for a pair of matrices are loops that are present or enriched in one matrix and not present or depleted in the other. An added practical benefit is that Mustache is fast enough that it can run on a CPU, whereas many other loop callers require a GPU.</div>
</div>
<section id="id5">
<h5>call_on<a class="headerlink" href="#id5" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>is_techrep <cite>default</cite></p></li>
<li><p>is_biorep <cite>default</cite></p></li>
<li><p>is_condition <cite>default</cite></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Whether to call loops on technical replicates, biological replicates, and conditions.</div>
</div>
</section>
<section id="use-format">
<h5>use_format<a class="headerlink" href="#use-format" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Options:</div>
</div>
<ul class="simple">
<li><p>mcool <cite>default</cite></p></li>
<li><p>hic</p></li>
</ul>
<div class="line-block">
<div class="line">Mustache can use both .mcool and .hic matrix formats as input. Loops will only be called on samples where the appropriate matrix type is output. If both are generated, which is used should not affect the outcome.</div>
</div>
</section>
<section id="mustache-params">
<h5>mustache_params<a class="headerlink" href="#mustache-params" title="Link to this heading">¶</a></h5>
<div class="line-block">
<div class="line">Default:</div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--resolution</span> <span class="pre">5000</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--pThreshold</span> <span class="pre">.1</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sparsityThreshold</span> <span class="pre">.88</span></code></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Parameters passed to mustache_diffloops, which will output both individual matrix loop calls and a pair of diffloops calls for each matrix.</div>
</div>
</section>
</section>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Hich</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Hich Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Hich Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sample-attributes">Sample attributes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#general">General</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-id">sample_id</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id">id</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly">assembly</a></li>
<li class="toctree-l4"><a class="reference internal" href="#datatype">datatype</a></li>
<li class="toctree-l4"><a class="reference internal" href="#is-techrep">is_techrep</a></li>
<li class="toctree-l4"><a class="reference internal" href="#is-biorep">is_biorep</a></li>
<li class="toctree-l4"><a class="reference internal" href="#is-condition">is_condition</a></li>
<li class="toctree-l4"><a class="reference internal" href="#min-mapq">min_mapq</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignment">Alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#aligner">aligner</a></li>
<li class="toctree-l4"><a class="reference internal" href="#aligner-threads">aligner_threads</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bwa-flags">bwa_flags</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#pairs-processing">Pairs processing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#deduplicate">deduplicate</a></li>
<li class="toctree-l4"><a class="reference internal" href="#pairs-format">pairs_format</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#chrom1">chrom1</a></li>
<li class="toctree-l5"><a class="reference internal" href="#pos1">pos1</a></li>
<li class="toctree-l5"><a class="reference internal" href="#chrom2">chrom2</a></li>
<li class="toctree-l5"><a class="reference internal" href="#pos2">pos2</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="#parse-params">parse_params</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dedup-params">dedup_params</a></li>
<li class="toctree-l4"><a class="reference internal" href="#select-params">select_params</a></li>
<li class="toctree-l4"><a class="reference internal" href="#select-condition">select_condition</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#keep-pair-types">keep_pair_types</a></li>
<li class="toctree-l5"><a class="reference internal" href="#keep-trans">keep_trans</a></li>
<li class="toctree-l5"><a class="reference internal" href="#keep-cis">keep_cis</a></li>
<li class="toctree-l5"><a class="reference internal" href="#min-dist-fr">min_dist_fr</a></li>
<li class="toctree-l5"><a class="reference internal" href="#min-dist-rf">min_dist_rf</a></li>
<li class="toctree-l5"><a class="reference internal" href="#min-dist-ff">min_dist_ff</a></li>
<li class="toctree-l5"><a class="reference internal" href="#id1">min_dist_ff</a></li>
<li class="toctree-l5"><a class="reference internal" href="#chroms">chroms</a></li>
<li class="toctree-l5"><a class="reference internal" href="#discard-same-frag">discard_same_frag</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#matrix-processing">Matrix processing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#make-hic">make_hic</a></li>
<li class="toctree-l4"><a class="reference internal" href="#make-cool">make_cool</a></li>
<li class="toctree-l4"><a class="reference internal" href="#make-mcool">make_mcool</a></li>
<li class="toctree-l4"><a class="reference internal" href="#matrix">matrix</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#make-mcool-file-format">make_mcool_file_format</a></li>
<li class="toctree-l5"><a class="reference internal" href="#make-hic-file-format">make_hic_file_format</a></li>
<li class="toctree-l5"><a class="reference internal" href="#resolutions">resolutions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#quality-control">Quality control</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicrep">hicrep</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#call-on">call_on</a></li>
<li class="toctree-l5"><a class="reference internal" href="#id2">resolutions</a></li>
<li class="toctree-l5"><a class="reference internal" href="#id3">chroms</a></li>
<li class="toctree-l5"><a class="reference internal" href="#exclude">exclude</a></li>
<li class="toctree-l5"><a class="reference internal" href="#chrom-filter">chrom_filter</a></li>
<li class="toctree-l5"><a class="reference internal" href="#h">h</a></li>
<li class="toctree-l5"><a class="reference internal" href="#dbpmax">dBPMax</a></li>
<li class="toctree-l5"><a class="reference internal" href="#bdownsample">bDownSample</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#feature-calling">Feature calling</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#compartments">compartments</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#resolution">resolution</a></li>
<li class="toctree-l5"><a class="reference internal" href="#cooltools-eigs-cis-params">cooltools_eigs_cis_params</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="#insulation">insulation</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#id4">resolution</a></li>
<li class="toctree-l5"><a class="reference internal" href="#cooltools-insulation-params">cooltools_insulation_params</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="#loops">loops</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#id5">call_on</a></li>
<li class="toctree-l5"><a class="reference internal" href="#use-format">use_format</a></li>
<li class="toctree-l5"><a class="reference internal" href="#mustache-params">mustache_params</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Hich Installation</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Benjamin Skubi, Andrew Adey, Galip Gürkan Yardımcı.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.0.2</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/reference.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>