

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Sample File &mdash; Hich 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=8ea2664f" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=2709fde1"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="_static/copybutton.js?v=f281be69"></script>
      <script>let toggleHintShow = 'Click to show';</script>
      <script>let toggleHintHide = 'Click to hide';</script>
      <script>let toggleOpenOnPrint = 'true';</script>
      <script src="_static/togglebutton.js?v=4a39c7ea"></script>
      <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
      <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Hich
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Install Nextflow and Apptainer</a></li>
<li class="toctree-l1"><a class="reference internal" href="features/features.html">Sample Selection Strategy</a></li>
<li class="toctree-l1"><a class="reference internal" href="vignettes/index.html">Vignettes</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Hich</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Sample File</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/basic_setup.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="sample-file">
<h1>Sample File<a class="headerlink" href="#sample-file" title="Link to this heading"></a></h1>
<p>The Hich sample file is a tab- or comma-delimited (.tsv or .csv) file that describes your experimental setup. It can often look as simple as this:</p>
<section id="example-1-multiple-conditions-bioreps-techreps-and-reference-genomes">
<h2>Example 1. Multiple conditions, bioreps, techreps, and reference genomes<a class="headerlink" href="#example-1-multiple-conditions-bioreps-techreps-and-reference-genomes" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>condition</p></th>
<th class="head"><p>biorep</p></th>
<th class="head"><p>techrep</p></th>
<th class="head"><p>fastq1</p></th>
<th class="head"><p>fastq2</p></th>
<th class="head"><p>assembly</p></th>
<th class="head"><p>enzymes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Human_KO</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>Human_KO_1_1_R1.fq.gz</p></td>
<td><p>Human_KO_1_1_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-odd"><td><p>Human_KO</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>Human_KO_1_2_R1.fq.gz</p></td>
<td><p>Human_KO_1_2_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-even"><td><p>Human_KO</p></td>
<td><p>2</p></td>
<td><p>1</p></td>
<td><p>Human_KO_2_1_R1.fq.gz</p></td>
<td><p>Human_KO_2_1_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-odd"><td><p>Human_KO</p></td>
<td><p>2</p></td>
<td><p>2</p></td>
<td><p>Human_KO_2_2_R1.fq.gz</p></td>
<td><p>Human_KO_2_2_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-even"><td><p>Mouse_KO</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>Mouse_KO_1_1_R1.fq.gz</p></td>
<td><p>Mouse_KO_1_1_R2.fq.gz</p></td>
<td><p>mm10</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-odd"><td><p>Mouse_KO</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>Mouse_KO_1_2_R1.fq.gz</p></td>
<td><p>Mouse_KO_1_2_R2.fq.gz</p></td>
<td><p>mm10</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-even"><td><p>Mouse_KO</p></td>
<td><p>2</p></td>
<td><p>1</p></td>
<td><p>Mouse_KO_2_1_R1.fq.gz</p></td>
<td><p>Mouse_KO_2_1_R2.fq.gz</p></td>
<td><p>mm10</p></td>
<td><p>Arima</p></td>
</tr>
<tr class="row-odd"><td><p>Mouse_KO</p></td>
<td><p>2</p></td>
<td><p>2</p></td>
<td><p>Mouse_KO_2_2_R1.fq.gz</p></td>
<td><p>Mouse_KO_2_2_R2.fq.gz</p></td>
<td><p>mm10</p></td>
<td><p>Arima</p></td>
</tr>
</tbody>
</table>
<p>The condition, biorep, and techrep columns label the condition, biological replicate, and technical replicate for each sample (row). The fastq1 and fastq2 columns give the path to the paired-end fastq files containing sequencing data for each sample. The assembly column contains the name of the reference genome (hg38 = human, mm10 = murine). The enzymes column can be a comma-separated list of restriction enzymes or the name of the kit used in the experiment.</p>
<p>With this setup, Hich will automatically download the reference genome (provided it recognizes the assembly name), indexing it for alignment with bwa-mem2, and producing other needed resource files such as a partition of the genome into restriction fragments. Note that resource files will be produced only once (so that here, one copy of the hg38 reference will be downloaded and shared by all samples).</p>
</section>
<section id="example-2-a-different-aligner">
<h2>Example 2. A different aligner<a class="headerlink" href="#example-2-a-different-aligner" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>condition</p></th>
<th class="head"><p>fastq1</p></th>
<th class="head"><p>fastq2</p></th>
<th class="head"><p>assembly</p></th>
<th class="head"><p>enzymes</p></th>
<th class="head"><p>aligner</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Human_KO</p></td>
<td><p>Human_KO_1_1_R1.fq.gz</p></td>
<td><p>Human_KO_1_1_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
<td><p>bwa</p></td>
</tr>
</tbody>
</table>
<p>Here, only one sample is being used. The biorep and techrep will default to “1”. There is also a new column, specifying that the “bwa” aligner should be used rather than the default “bwa-mem2”. While the default bwa-mem2 aligner is faster, the alternative bwa aligner requires less memory for indexing, which is the main motivation for offering it as an option. Both aligners should produce identical outputs.</p>
</section>
<section id="example-3-use-local-reference-index-chromsizes-and-fragfile">
<h2>Example 3. Use local reference, index, chromsizes, and fragfile<a class="headerlink" href="#example-3-use-local-reference-index-chromsizes-and-fragfile" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>condition</p></th>
<th class="head"><p>fastq1</p></th>
<th class="head"><p>fastq2</p></th>
<th class="head"><p>assembly</p></th>
<th class="head"><p>enzymes</p></th>
<th class="head"><p>aligner</p></th>
<th class="head"><p>reference</p></th>
<th class="head"><p>chromsizes</p></th>
<th class="head"><p>index_dir</p></th>
<th class="head"><p>index_prefix</p></th>
<th class="head"><p>fragfile</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Human_KO</p></td>
<td><p>Human_KO_1_1_R1.fq.gz</p></td>
<td><p>Human_KO_1_1_R2.fq.gz</p></td>
<td><p>hg38</p></td>
<td><p>Arima</p></td>
<td><p>bwa</p></td>
<td><p>resources/hg38.fa</p></td>
<td><p>resources/hg38.sizes</p></td>
<td><p>resources/bwa</p></td>
<td><p>GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta</p></td>
<td><p>resources/hg38_Arima.bed</p></td>
</tr>
</tbody>
</table>
<p>Here, all resource files are already created and their paths are given, but this is not necessary. The <code class="docutils literal notranslate"><span class="pre">index_dir</span></code> and <code class="docutils literal notranslate"><span class="pre">index_prefix</span></code> columns must either both have a value or neither have a value, but other than that, it is fine to specify some resource file columns and leave others blank for any given sample. The new columns are:</p>
<ul class="simple">
<li><p>reference: The reference genome in FASTA format</p></li>
<li><p>chromsizes: A two-column tab-delimited file with the contigs in column 1 and the size (bp) of the contig in column 2.</p></li>
<li><p>index_dir: The directory where the desired index files are located</p></li>
<li><p>index_prefix: The common prefix string for each index file. BWA, for example, uses the files [index_prefix].amb, [index_prefix].ann, [index_prefix].bwt, [index_prefix].pac, and [index_prefix].sa.</p></li>
<li><p>fragfile: A partition of the genome into restriction fragments in BED format.</p></li>
</ul>
<p>4D Nucleome makes several commonly-used BWA index files available for manual download. If using them, you will need to download and unzip them before putting their paths into the sample file. Many labs produce an index for the genomes they work with and reuse them across many projects. As indexing the genome is the most RAM-intensive process Hich executes, it can be helpful to supply one if available.</p>
<p>Links to 4DN BWA Index files:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFIUH46PG1/#details">danRer11 (compressed)</a></p></li>
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFIZ2PWCC2/#details">mm10 (uncompressed)</a></p></li>
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFIZQB369V/#details">hg38 (uncompressed)</a></p></li>
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFIVGRYVQF/#details">galGal5 (compressed)</a></p></li>
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFIO5MGY32/#details">dm6 (compressed)</a></p></li>
<li><p><a class="reference external" href="https://data.4dnucleome.org/files-reference/4DNFI823LSI8/#details">mm10 (compressed)</a></p></li>
</ul>
<p>If using one of these, it is essential to set the <code class="docutils literal notranslate"><span class="pre">aligner</span></code> column to <code class="docutils literal notranslate"><span class="pre">bwa</span></code> in the sample file for each sample using a BWA index, since Hich defaults to using the faster (but more RAM-intensive) bwa-mem2, which uses a different index format.</p>
</section>
<section id="merging-and-downsampling">
<h2>Merging and downsampling<a class="headerlink" href="#merging-and-downsampling" title="Link to this heading"></a></h2>
<p>In this documentation, some terms have specific definitions.</p>
<ul class="simple">
<li><p><strong>Sample</strong> A sample is the complete collection of data and parameters associated with a single technical replicate (techrep), biological replicate (biorep), or condition.</p></li>
<li><p><strong>Biological sample (biosample)</strong> A biosample refers to the physical material (cells, a sequencing library prep) that is associated with a particular Hich sample.</p></li>
<li><p><strong>Condition</strong> An experimental condition, such as “No treatment (NT)” or “Knockout (KO)” or data obtained from different cell lines or drug exposures. Conditions are formed by merging their respective bioreps, but are not deduplicated as their input biorep samples will have already been deduplicated.</p></li>
<li><p><strong>Biological replicate</strong> Each condition can have one or more biological replicates, representing data from the same experimental condition applied to one or more biosamples. Bioreps are formed by merging their respective techreps, and are then deduplicated before contact matrices are generated.</p></li>
<li><p><strong>Technical replicate</strong> Each biorep can have one or more technical replicates, representing data from the same biorep sequenced multiple times. After merging techreps to form bioreps (which are independently deduplicated), the individual techreps are deduplicated as well.</p></li>
<li><p><strong>Input sample</strong> A sample that is input to Hich as a row in the sample file</p></li>
<li><p><strong>Synthetic sample</strong> A sample that is created by Hich based on input samples, typically via merging and/or downsampling.</p></li>
</ul>
<p>When Hich merges techrep samples to form biorep samples, it keeps the techrep samples as well. It also keeps the biorep samples after merging them to form condition samples. That means that the output contains techrep, biorep, and condition samples for the experiment, facilitating downstream QC and analysis.</p>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Benjamin Skubi, Andrew Adey, Galip Gürkan Yardımcı.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>